src/Parser.ts

```diff
@@ -0,0 +1,218 @@
+import Program from './AST/Program';
+
+enum Precedence {
+  lowest,
+  index,
+}
+
+type PrefixParser = () => ASTExpression | undefined;
+
+type InfixParser = (arg?: ASTExpression) => ASTExpression | undefined;
+
+interface IParser<ParserType> {
+  [key: symbol]: ParserType;
+}
+
+class Parser {
+  private lexer: Lexer;
+  private errors: Array<string> = [];
+
+  private curToken!: Token;
+  private peekToken!: Token;
+
+  private prefixParseFuncs: IParser<PrefixParser>;
+  private infixParseFuncs: IParser<InfixParser>;
+
+  private precedenceMap: { [key: symbol]: Precedence } = {
+    [lbracket]: Precedence.index,
+  };
+
+  constructor(lexer: Lexer) {
+    this.lexer = lexer;
+    this.prefixParseFuncs = {};
+    this.infixParseFuncs = {};
+
+    this.registerPrefix(this.parseSelectExpression.bind(this), dot);
+    this.registerPrefix(this.parseIdentifier.bind(this), ident);
+    this.registerPrefix(this.parseIntegerLiteral.bind(this), int);
+
+    this.registerInfix(this.parseIndexExpression.bind(this), lbracket);
+
+    this.nextToken();
+    this.nextToken();
+  }
+
+  public parseProgram(): Program {
+    let program = new Program();
+
+    while (!this.isCurToken(eof)) {
+      const stmt = this.parseStatement();
+      if (!stmt) {
+        this.nextToken();
+        continue;
+      }
+
+      program.statements.push(stmt);
+
+      this.nextToken();
+    }
+
+    if (this.errors.length > 0) {
+      throw this.errors;
+    }
+
+    return program;
+  }
+
+  // -- Token Handling --
+
+  private nextToken(): void {
+    this.curToken = this.peekToken;
+    this.peekToken = this.lexer.nextToken();
+  }
+
+  private isCurToken(type: symbol): boolean {
+    return this.curToken.type === type;
+  }
+
+  private isPeekToken(type: symbol): boolean {
+    return this.peekToken.type === type;
+  }
+
+  private expectPeek(type: symbol): boolean {
+    if (this.isPeekToken(type)) {
+      this.nextToken();
+
+      return true;
+    } else {
+      this.peekError(type);
+
+      return false;
+    }
+  }
+
+  private peekError(type: symbol): void {
+    this.errors.push(
+      `expected next token to be ${type.toString()}, got ${this.peekToken.type.toString()} instead`,
+    );
+  }
+
+  private noPrefixParseFuncErr(type: String): void {
+    this.errors.push(`prefix parse func for ${type} not found`);
+  }
+
+  // -- Precedence --
+
+  private peekPrecedence(): Precedence {
+    const precedence = this.precedenceMap[this.peekToken.type];
+    if (!precedence) {
+      return Precedence.lowest;
+    }
+
+    return precedence;
+  }
+
+  // -- Parsing --
+
+  // JSL only has one type of statement right now.
+  private parseStatement(): ASTStatement {
+    return this.parseExpressionStatement();
+  }
+
+  private parseExpressionStatement(): ASTExpressionStatement {
+    return new ASTExpressionStatement(this.curToken, this.parseExpression(Precedence.lowest));
+  }
+
+  private parseExpression(precedence: Precedence): ASTExpression | undefined {
```

Comment:
agree with you!

Reply:
If we are going to have `ASTExpression | null` as the return type for this method, then it would throw an error on other functions that make use of the returned value from this method. As we also have the ASTs which support only a particular type of ASTNode or undefined, it's not possible for us to pass a null value to it. We can do it but should add couple of checks to validate that we don't send a null to ASTs which don't support it.